tenantcostclient: include all requests in eCPU model

andy-kimball opened the pull request:
The RU model only included a subset of batch request types in its cost
calculations. Update the estimated CPU model to include all types, in
order to increase its accuracy. Doing this requires updating the
OnResponseWait method to take a BatchRequest, BatchResponse, and target
range and replica descriptors, rather than RequestInfo and ResponseInfo.
This allows the tenant controller to decide which request types to
include, based on the model that's currently in use.

Epic: https://cockroachlabs.atlassian.net/browse/CC-28471

Release note: None

andy-kimball committed:
tenantcostclient: move code from kvcoord to tenantcostclient

Move the computeNetworkCost and computeWriteReplicationNetworkPaths functions
from the kvcoord package to the tenantcostclient package. This is part of a
larger change to refactor the OnResponseWait function on the tenant-side cost
controller. Keeping it in a separate commit makes the larger change easier to
review.

Epic: https://cockroachlabs.atlassian.net/browse/CC-28471

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127142)
<!-- Reviewable:end -->

jaylim-crl commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go:
```diff
@@ -121,14 +122,27 @@ const bufferTokens = 5000
 // magnitude as request units.
 const tokensPerCPUSecond = 1000
 
+// LocalityNetworkPath describes a source -> destination network flow.
+type LocalityNetworkPath struct {
```
This is no longer used, now that we're incrementing the metric directly.

andy-kimball replied:
Removed.

jaylim-crl commented on pkg/ccl/multitenantccl/tenantcostclient/testdata/consumption:
```diff
@@ -732,31 +700,28 @@ Estimated CPU seconds: 0.00
-# Test multiple requests in the same batch with RU multiplier of 1.
-write count=2 bytes=1024 networkCost=0
+# Test multiple requests in the same batch.
+write count=2 bytes=1024 localities=cross-region
```
Hm, do you know why `estimated_replication_bytes` is still 0 despite doing a cross-region write?

andy-kimball replied:
Because we're only computing `estimated_replication_bytes` when the eCPU model is enabled. When the RU model is enabled, we instead compute `tenant.sql_usage.cross_region_network_ru`. I figured we wouldn't want to compute `estimated_replication_bytes` for Basic clusters.

jaylim-crl commented on pkg/kv/kvserver/replica_rate_limit.go:
```diff
@@ -32,9 +32,20 @@ func (r *Replica) maybeRateLimitBatch(ctx context.Context, ba *kvpb.BatchRequest
+	var info tenantcostmodel.BatchInfo
+	for i := range ba.Requests {
+		req := ba.Requests[i].GetInner()
+		if !kvpb.IsReadOnly(req) {
```
The RU model only includes a subset of batch requests, whereas here, we're counting everything, which is desired outcome for eCPU models. What if the tenant is still using the RU model (but the limiter code here has already been updated for the eCPU model)?

andy-kimball replied:
This is the KV tenant rate limiter, which is not related to the customer's model. It exists to prevent each KV node from being overloaded by a single tenant. There's a separate issue to convert this code to use the eCPU model, since it's more accurate. However, even before that happens, it's OK to loosen this code up a bit to include all requests, since none of this impacts customer costs.

jaylim-crl reviewed (approved):
LGTM

jeffswenson reviewed (approved):
LGTM

andy-kimball committed:
tenantcostclient: include all requests in eCPU model

The RU model only included a subset of batch request types in its cost
calculations. Update the estimated CPU model to include all types, in
order to increase its accuracy. Doing this requires updating the
OnResponseWait method to take a BatchRequest, BatchResponse, and target
range and replica descriptors, rather than RequestInfo and ResponseInfo.
This allows the tenant controller to decide which request types to
include, based on the model that's currently in use.

Epic: https://cockroachlabs.atlassian.net/browse/CC-28471

Release note: None
